### PR TITLE
Decryptor selection (please do not merge yet)

### PR DESCRIPTION
**Pull request by calvaris**
As the title says, please do not merge yet, I can push it myself cause I am doing the final test. The PR is to allow comments.

What this branch does is taking advantage of the patch for https://bugzilla.gnome.org/show_bug.cgi?id=770107 and respond to the need-context message. The implemented answer is compliant with EME 0.1b and the non promise based object oriented version (aka V2).

It basically answers the query, blocks the thread and waits for some key request generation. This causes some issues that are addressed in the rest of the patches.

### CODE REVIEW

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 1]
```diff
@@ -313,68 +324,134 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
...
 #else
-            ASSERT_NOT_REACHED();
+                ASSERT_NOT_REACHED();
 #endif
-            gst_buffer_unmap(data, &mapInfo);
+            });
+
+            GST_INFO("waiting for a key request to arrive");
+            LockHolder lock(m_protectionMutex);
+            m_protectionCondition.waitForSeconds(m_protectionMutex, 4);
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Maybe try with this? It will block for 4 seconds max, if the predicate returns true it will unblock before the timeout.

m_protectionCondition.waitFor(m_protectionMutex, std::chrono::seconds(4), [this] { return this->m_lastGeneratedKeyRequestKeySystemUuid.isEmpty(); });

File: Source/WTF/wtf/Condition.h [hunk 2]
```diff
@@ -142,6 +142,12 @@ struct ConditionBase {
         return waitForSecondsImpl(lock, absoluteTimeoutSeconds - monotonicallyIncreasingTime());
     }
     
+    template<typename LockType>
+    bool waitForSeconds(LockType& lock, double relativeTimeoutSeconds)
```
Comment (philn) [inline Source/WTF/wtf/Condition.h]: Why is this needed? I propose a different approach, see down below.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 3]
```diff
@@ -1291,19 +1369,49 @@ MediaPlayer::MediaKeyException MediaPlayerPrivateGStreamerBase::addKey(const Str
...
 }
 
+void MediaPlayerPrivateGStreamerBase::trimInitData(String keySystemUuid, const unsigned char*& initDataPtr, unsigned &initDataLength)
+{
+    if (!m_initDatas.contains(keySystemUuid)) {
+        GST_TRACE("we don't have an initData for %s", keySystemUuid.utf8().data());
+        return;
+    }
+
+    Vector<uint8_t> storedInitData = m_initDatas.get(keySystemUuid);
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Uint8Array ?

File: Source/WebCore/platform/graphics/gstreamer/WebKitCommonEncryptionDecryptorGStreamer.cpp [hunk 4]
```diff
@@ -296,33 +294,19 @@ static gboolean webkitMediaCommonEncryptionDecryptSinkEventHandler(GstBaseTransf
...
-                    gst_event_unref(self->priv->protectionEvent);
-                    self->priv->protectionEvent = nullptr;
-                }
-            }});
+        gst_event_parse_protection(event, &systemId, nullptr, nullptr);
+        GST_TRACE_OBJECT(self, "received protection event for %s", systemId);
 
+        if (g_strcmp0(systemId, klass->protectionSystemId) == 0) {
+            GST_DEBUG_OBJECT(self, "sending protection event to the pipeline");
+            gst_element_post_message(GST_ELEMENT(self),
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/WebKitCommonEncryptionDecryptorGStreamer.cpp]: This should be done from the main thread, like it used to be.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp (hunk 1 above)
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: I will try it

File: Source/WTF/wtf/Condition.h (hunk 2 above)
Comment (calvaris) [inline Source/WTF/wtf/Condition.h]: This is needed because we cannot let the qtdemux select the decryptor without us checking with the JS app and therefore it needs to be done here and wait for the result. Otherwise, when we got the answer, we might have the wrong decryptor selected or even worse, if we just let the decryptors filter the events, we might be screwed cause we wouldn't be getting all of them, which is needed for 0.1b.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp (hunk 3 above)
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: I can try it but I need so ensure I can use this to search on the contained data easily.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp (hunk 1 above)
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Yes, it works, when I read the code at the beginning I had the impression that it would wait forever but I was clearly wrong and actually this works much better.

Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: It creates a regression I have to investigate.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp (hunk 3 above)
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: The problem with this is that it is not easy so add and remove a character so I prefer to stick to Vector.
